src/conn/conn_compact.c

```diff
@@ -18,6 +21,63 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
     return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
 }
 
+/*
+ * __find_next_uri --
+ *     Given a URI, find the next one in the metadata file that is eligible for compaction.
+ */
+static int
+__find_next_uri(WT_SESSION_IMPL *session, const char *uri, const char **next_uri)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    int exact;
+    const char *key;
+
+    cursor = NULL;
+    key = NULL;
+    exact = 0;
+
+    /* Use a metadata cursor to have access to the existing URIs. */
+    WT_ERR(__wt_metadata_cursor(session, &cursor));
+
+    /* Position the cursor on the given URI. */
+    cursor->set_key(cursor, uri);
+    WT_ERR(cursor->search_near(cursor, &exact));
+
+    /*
+     * The given URI may not exist in the metadata file. Since we always want to return a URI that
+     * is lexicographically larger the given one, make sure not to go backwards.
+     */
+    if (exact <= 0)
+        WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+
+    /* Loop through the eligible candidates. */
+    while (ret == 0) {
+        WT_ERR(cursor->get_key(cursor, &key));
+        /* Check we are still dealing with keys which have the right prefix. */
+        if (!WT_PREFIX_MATCH(key, WT_BG_COMPACT_URI_PREFIX)) {
+            ret = WT_NOTFOUND;
+            break;
+        }
+        /* There are files that should not be compacted. */
+        if (!WT_STREQ(key, WT_HS_URI))
+            /* FIXME-WT-11343: check if the table is supposed to be compacted. */
+            break;
+        WT_ERR_NOTFOUND_OK(cursor->next(cursor), true);
+    }
+
+    /* Save the selected uri. */
+    if (ret == 0) {
+        __wt_free(session, *next_uri);
```

Comment:
I'm not crazy about calling free here. This function doesn't know what `*next_uri` might point to or whether it is safe to free it. It could point to something valid in the caller. I.e., suppose the caller does something like:
```
    save_uri = next_uri;
    __find_next_uri(..., &next_uri);
```

This code would wind up freeing memory the caller is still using.

In fact, I would initialize `*next_urip = NULL` at the beginning of this function. That would ensure that it is set even in error returns. And if the caller forgets to check the return value and tries to use `*next_urip`, it will get a core dump and a very visible pointer to the problem.

Reply:
I see, I have initialized that value as suggested in 0b262a95fc4e8ad5656e1c4b170a39229019e245, thanks!